File: p085.java
/* 
 * Solution to Project Euler problem 85
 * By Nayuki Minase
 * 
 * http://nayuki.eigenstate.org/page/project-euler-solutions
 * https://github.com/nayuki/Project-Euler-solutions
 */


public final class p085 implements EulerSolution {
	
	public static void main(String[] args) {
		System.out.println(new p085().run());
	}
	
	
	public String run() {
		int bestDiff = -1;
		int bestArea = -1;
		for (int w = 1; w < 2000; w++) {
			for (int h = 1; h < 2000; h++) {
				int diff = Math.abs(numberOfRectangles(w, h) - 2000000);
				if (bestDiff == -1 || diff < bestDiff) {
					bestDiff = diff;
					bestArea = w * h;
				}
			}
		}
		return Integer.toString(bestArea);
	}
	
	
	private static int numberOfRectangles(int m, int n) {
		return (m + 1) * m * (n + 1) * n / 4;
	}
	
}

File: p085.py
# 
# Solution to Project Euler problem 85
# by Project Nayuki
# 
# http://www.nayuki.io/page/project-euler-solutions
# https://github.com/nayuki/Project-Euler-solutions
# 

import eulerlib


def compute():
	TARGET = 2000000
	bestdiff = None
	end = eulerlib.sqrt(TARGET) + 1
	for w in range(1, end):
		for h in range(1, end):
			diff = abs(num_rectangles(w, h) - TARGET)
			if bestdiff is None or diff < bestdiff:
				ans = w * h
				bestdiff = diff
	return str(ans)


def num_rectangles(m, n):
	return (m + 1) * m * (n + 1) * n // 4  # A bit more than m^2 n^2 / 4


if __name__ == "__main__":
	print(compute())
